### PR TITLE
Proposed implementation for POEM060 - reports system

### LINKED ISSUE #2411: Create reference implementation of POEM 060 - HTML Report Generation #131
### Issue Type

- [ ] Bug
- [x] Enhancement  
      POEM 060

### Description

Create a reference implementation for 

[https://github.com/OpenMDAO/POEMs/pull/131](https://github.com/OpenMDAO/POEMs/pull/131)

### Example

N/A

### Environment

N/A

### PR DESCRIPTION
**Pull request by hschilling**
### Summary

Added code to automatically generate reports when running problems by default. Now, by default, the N2 diagram and the scaling report are generated automatically and placed in a directory whose name is based on the Problem name.

This is the proposed implementation for the POEM

[https://github.com/OpenMDAO/POEMs/blob/bee7fe90a93ef0ce12d37dca95e7b58401a46e6e/POEM_060.md](https://github.com/OpenMDAO/POEMs/blob/bee7fe90a93ef0ce12d37dca95e7b58401a46e6e/POEM_060.md)

### Related Issues

- Resolves #2411 

### Backwards incompatibilities

None

### New Dependencies

None

### CODE REVIEW

**Comment by robfalck** [comment]
Based on experience in running a lot of subproblems, the report directory can get cluttered up pretty easily.  To address this:

1. Make `./reports` the default report directory.  Users can change to `./` if they prefer, but this default reduces the potential to really clutter up your working directory.

2. Add a new argument to `Problem.__init__`

- `reports` defaults to `_unspecified`, and if unspecified the `OPENMDAO_REPORTS` variable is used.
- If given, `reports` should override `OPENMDAO_REPORTS`.  If boolean, enable/disable all reports.  Since `none` is acceptable in the environment variable, a value of `reports=None` should be equivalent to `reports=False`.  Otherwise, `reports` may be a sequence of strings giving the names of the reports to run.

**Review by robfalck — CHANGES_REQUESTED** [review]
See my comments on PR thread, but I think we should add arguments `reports` and `reports_dir` to `Problem`.  These would override the corresponding environment variables if provided (default to unspecified).

File: openmdao/docs/openmdao_book/features/model_visualization/reports_system.ipynb [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+   "execution_count": null,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "import pathlib\n",
+    "from openmdao.utils.reports_system import register_report, get_reports_dir\n",
```
Comment (robfalck) [inline openmdao/docs/openmdao_book/features/model_visualization/reports_system.ipynb:127]: `register_reports` and `get_reports_dir` should be imported from `openmdao.api` in the documentation, to simplify things for users.

File: openmdao/core/problem.py [hunk 2]
```diff
@@ -137,9 +143,16 @@ class Problem(object):
...
+        If _UNDEFINED, the OPENMDAO_REPORTS variable is used. Defaults to _UNDEFINED.
+        If given, reports should override OPENMDAO_REPORTS. If boolean, enable/disable all reports.
+        Since none is acceptable in the environment variable, a value of reports=None
+        is equivalent to reports=False. Otherwise, reports may be a sequence of
+        strings giving the names of the reports to run.
     """
 
-    def __init__(self, model=None, driver=None, comm=None, name=None, **options):
+    def __init__(self, model=None, driver=None, comm=None, name=None,
+                 reports=_UNDEFINED, **options):
```
Comment (robfalck) [inline openmdao/core/problem.py:155]: Add `reports_dir` here with similar behavior to `reports`, such that it defaults to _UNDEFINED and overrides the default/environment variable if provided.

File: openmdao/docs/openmdao_book/features/model_visualization/reports_system.ipynb [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    "import openmdao.api as om\n",
+    "from openmdao.test_suite.components.paraboloid import Paraboloid\n",
+    "\n",
+    "user_report_filename = 'user_report.txt'\n",
+    "\n",
+    "def user_defined_report(prob):\n",
```
Comment (robfalck) [inline openmdao/docs/openmdao_book/features/model_visualization/reports_system.ipynb:133]: This is a lot of boilerplate to work out the file to open/write.  Should we provide `problem_reports_dirpath` as the second argument of the report generation function?

```
def user_defined_report(prob, report_dir_path):
...
```

**Review by robfalck — CHANGES_REQUESTED** [review]

File: openmdao/docs/openmdao_book/features/model_visualization/reports_system.ipynb [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Reports System\n",
+    "\n",
+    "## Default Usage\n",
+    "\n",
+    "OpenMDAO has a reports system which will generate several common reports when you run your model. Currently, the only reports that are run are the N2 diagram and the scaling report. (The scaling report does not work with MPI currently.) The names of these reports in the reporting system, are \"n2\" and \"scaling\", respectively.\n",
```
Comment (naylor-b) [inline openmdao/docs/openmdao_book/features/model_visualization/reports_system.ipynb:29]: Can remove "the scaling report does not work with MPI ..."

File: openmdao/utils/reports_system.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+# Keeping track of the registered reports
+_Report = namedtuple('Report', 'func desc class_name inst_id method pre_or_post')
+_reports_registry = {}
+
+_reports_dir = './reports'  # the default location for the reports
+
+
+def _is_rank_0(prob):
+    # Utility function to determine if MPI and on rank0 or not on MPI at all
+    on_rank0 = True
```
Comment (naylor-b) [inline openmdao/utils/reports_system.py:28]: Could simplify this whole function to `return not MPI or prob.comm.rank == 0`

**Review by naylor-b — CHANGES_REQUESTED** [review]

File: openmdao/utils/reports_system.py (hunk 5 above)
Comment (hschilling) [reply openmdao/utils/reports_system.py:28]: Great idea. Done

**Review by naylor-b — CHANGES_REQUESTED** [review]

**Review by Kenneth-T-Moore — APPROVED** [review]
I wonder if we need some additional explanation in the docs for each report. I guess each report is covered elsewhere, if they put the dots together.

**Comment by hschilling** [comment]
> 

Good idea! I did add some info about the reports to the docs. No need for the user to do extra work looking for info elsewhere.

File: openmdao/docs/openmdao_book/features/model_visualization/reports_system.ipynb [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    "\n",
+    "The user can also have more fine-grained control over what reports get run by setting this environment variable to a string that lists the reports to run. For example\n",
+    "\n",
+    "1. OPENMDAO_REPORTS=n2\n",
+    "\n",
+    "2. OPENMDAO_REPORTS=n2,scaling\n",
+    "\n",
+    "3. OPENMDAO_REPORTS=none\n",
+    "\n",
+    "If `OPENMDAO_REPORTS` is not set, all reports are run. The user can also explicitly turn on all reports by setting `OPENMDAO_REPORTS` to any of these four values: '1', 'true', 'on', or 'all'.\n",
```
[the remark this answers is not in the document]
Comment (hschilling) [reply openmdao/docs/openmdao_book/features/model_visualization/reports_system.ipynb:57]: Great point. Fixed

File: openmdao/docs/openmdao_book/features/model_visualization/reports_system.ipynb [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Setting the directory for the reports\n",
+    "\n",
+    "If the user wishes to have the reports directory placed into a different directory other than the default current working\n",
+    "directory, the user can set the environment variable, `OPENMDAO_REPORTS_DIR` to the name of that directory. The\n",
+    "directory does not have to exist beforehand. It will be created by the reporting system. As an example, if `OPENMDAO_REPORTS_DIR` is set to \"my_reports\", then, if the `Problem` name is `problem1`, then the reports will be written to the \"my_reports/problem1\" directory.\n",
```
[the remark this answers is not in the document]
Comment (hschilling) [reply openmdao/docs/openmdao_book/features/model_visualization/reports_system.ipynb:78]: Good catch! Fixed

File: openmdao/utils/reports_system.py [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+                val = name
+            else:
+                val = getattr(report, column_name)
+                if column_name == 'func':
+                    val = val.__name__
+                else:
+                    val = str(val)
+            column_widths[column_name] = max(column_widths[column_name], len(val))
+
+    out_stream.write("\nHere are the reports registered to run:\n\n")
```
[the remark this answers is not in the document]
Comment (hschilling) [reply openmdao/utils/reports_system.py:143]: Wording changed.

**Review by naylor-b — APPROVED** [review]

**Pull request merged**